### PR TITLE
Test: Write the unit test suite

### PR DESCRIPTION
**Pull request by Hywan**
Related to https://github.com/hoaproject/Websocket/issues/64.

### Progression

* [x] `Client`,
* [x] `Connection`,
* [x] `Exception`:
  * [x] `BadProtocol`,
  * [x] `CloseError`,
  * [x] `Exception`,
  * [x] `InvalidMessage`,
* [x] `Node`,
* [x] `Protocol`:
  * [x] `Generic`,
  * [x] `Hybi00`,
  * [x] `Rfc6455`,
* [x] `Server`,
* [x] `Socket`.

### CODE REVIEW

**Comment by Hywan** [comment]
```
Success (8 test suites, 63/64 test cases, 0 void test case, 1 skipped test case, 287 assertions)!
```

Good progresses. The hardest part of the RFC6455 protocol is written.

**Comment by Hywan** [comment]
```sh
Success (1 test suite, 40/40 test cases, 0 void test case, 0 skipped test case, 680 assertions)!
```

**Comment by Hywan** [comment]
```sh
Success (10 test suites, 89/89 test cases, 0 void test case, 0 skipped test case, 940 assertions)!
```

**Comment by Hywan** [comment]
Left to do: `Connection`, `Client` and `Server`. Nice!

**Comment by Hywan** [comment]
```sh
Success (11 test suites, 97/97 test cases, 0 void test case, 0 skipped test case, 982 assertions)!
```
Better and better.

**Comment by Hywan** [comment]
Almost there!
```
Success (12 test suites, 147/147 test cases, 0 void test case, 0 skipped test case, 3484 assertions)!
```

**Comment by Hywan** [comment]
Almost there!
```
Success (12 test suites, 147/147 test cases, 0 void test case, 0 skipped test case, 3484 assertions)!
```

**Comment by Hywan** [comment]
```
Success (12 test suites, 151/151 test cases, 0 void test case, 0 skipped test case, 3514 assertions)!
```
and it's over!

File: Test/Unit/Client.php [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+                $node   = new \Mock\Hoa\Websocket\Node(),
+                $this->mockGenerator->makeVisible('_run')->generate(SUT::class),
+                $client = new \Mock\Hoa\Websocket\Client($socket),
+
+                $this->calling($node)->isMessageComplete = true,
+                $this->calling($socket)->getCurrentNode  = $node,
+                $this->calling($socket)->isDisconnected  = false,
+
+                $this->calling($client)->_run = function (Websocket\Node $_node) use (&$called, $self, $node) {
+                    $called = true;
```
Comment (jubianchi) [inline Test/Unit/Client.php]: you should assert on mock calls instead of working with this bool

File: Test/Unit/Client.php [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+    }
+
+    public function case_close()
+    {
+        return $this->_case_close(true);
+    }
+
+    public function case_close_with_no_protocol_implementation()
+    {
+        return $this->_case_close(null);
```
Comment (jubianchi) [inline Test/Unit/Client.php]: such things should be done with a data provider and an `->assert`

File: Test/Unit/Connection.php [hunk 3]
```diff
@@ -0,0 +1,1333 @@
...
+                ],
+                $this->calling($protocol)->writeFrame = function ($_message, $_opcode, $_mask) use (&$calledA) {
+                    $calledA = true;
+
+                    return;
+                },
+                $this->calling($connection)->close = function ($_code, $_reason) use (&$calledB, $self) {
+                    $calledB = true;
+
+                    $self
```
Comment (jubianchi) [inline Test/Unit/Connection.php]: using mock assertions (`->mock()->call()->withArguments()->exactly()`) you can avoid the ˋ$self` and manual assertions in stubs

File: Test/Unit/Connection.php [hunk 4]
```diff
@@ -0,0 +1,1333 @@
...
+        $self = $this;
+
+        $this
+            ->given(
+                $socket     = new \Mock\Hoa\Socket\Client('tcp://*:1234'),
+                $connection = new SUT($socket),
+                $protocol   = new \Mock\Hoa\Websocket\Protocol\Rfc6455($socket),
+                $this->mockGenerator->orphanize('__construct'),
+                $node = new \Mock\Hoa\Websocket\Node(),
+                $node->setProtocolImplementation($protocol),
```
Comment (jubianchi) [inline Test/Unit/Connection.php]: orphanizing constructor and using actual methods is risky

File: Test/Unit/Exception/BadProtocol.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * @license    New BSD License
+ */
+class BadProtocol extends Test\Unit\Suite
+{
+    public function case_hoa_websocket_exception()
+    {
+        $this
+            ->when($result = new SUT('foo', 0))
+            ->then
+                ->object($result)
```
Comment (jubianchi) [inline Test/Unit/Exception/BadProtocol.php]: s/object/exception/

File: Test/Unit/Exception/BadProtocol.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                    ->isInstanceOf(Websocket\Exception::class);
```
Comment (jubianchi) [inline Test/Unit/Exception/BadProtocol.php]: this should be a test on the class, not on an instance using ->class(SUT)->extends(...)

File: Test/Unit/Node.php [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+ *
+ * @copyright  Copyright © 2007-2016 Hoa community
+ * @license    New BSD License
+ */
+class Node extends Test\Unit\Suite
+{
+    public function case_is_a_node()
+    {
+        $this
+            ->given($this->mockGenerator->orphanize('__construct'))
```
Comment (jubianchi) [inline Test/Unit/Node.php]: no need to orphanize here if you test the class instead of an instance

File: Test/Unit/Client.php (hunk 2 above)
Comment (Hywan) [inline Test/Unit/Client.php]: Yup but data provider counts only 1 test case, and I prefer explicit naming too.

File: Test/Unit/Connection.php (hunk 4 above)
Comment (Hywan) [inline Test/Unit/Connection.php]: Yup :-(. Testability is not ideal here.

File: Test/Unit/Exception/BadProtocol.php (hunk 6 above)
Comment (Hywan) [inline Test/Unit/Exception/BadProtocol.php]: Oh, we have the `class` asserter group? Can you make a PR to update them?

**Pull request merged**